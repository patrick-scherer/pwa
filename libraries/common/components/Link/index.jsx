import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { openPageExtern } from '@shopgate/pwa-core';
import connect from './connector';
import styles from './style';

/**
 * Link component.
 * @param {Object} props Props for the component.
 * @returns {JSX}
 */
class Link extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    historyPush: PropTypes.func.isRequired,
    historyReplace: PropTypes.func.isRequired,
    href: PropTypes.string.isRequired,
    'aria-hidden': PropTypes.bool,
    'aria-label': PropTypes.string,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    openExtern: PropTypes.bool,
    replace: PropTypes.bool,
    role: PropTypes.string,
    state: PropTypes.shape(),
    tabIndex: PropTypes.number,
    tag: PropTypes.string,
  };

  static defaultProps = {
    'aria-hidden': null,
    'aria-label': null,
    className: '',
    disabled: false,
    replace: false,
    role: 'link',
    tag: 'div',
    tabIndex: null,
    openExtern: false,
    state: {},
  };

  /**
   * Opens the link.
   */
  handleOpenLink = () => {
    if (this.props.disabled) {
      return;
    }

    /**
     * For external pages we need to use the app command
     * iOS deeplinks work out of the box on
    * android though just opens it in the webview no matter
     * what. Also our router removes trailing slashes which
     * might be needed for external urls.
     */
    if (this.props.openExtern) {
      openPageExtern({ src: this.props.href });
      return;
    }

    const params = {
      pathname: this.props.href,
      state: this.props.state || {},
    };

    if (this.props.replace) {
      this.props.historyReplace(params);
    } else {
      this.props.historyPush(params);
    }
  };

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    const {
      tag: Tag, className, href, children, role, 'aria-label': ariaLabel, 'aria-hidden': ariaHidden, tabIndex,
    } = this.props;
    return (
      <Tag
        className={`${styles} ${className}`}
        onClick={this.handleOpenLink}
        role={role}
        data-test-id={`link: ${href}`}
        aria-label={ariaLabel}
        tabIndex={tabIndex}
        aria-hidden={ariaHidden}
      >
        {children}
      </Tag>
    );
  }
}

export const Disconnected = Link;

export default connect(Link);
