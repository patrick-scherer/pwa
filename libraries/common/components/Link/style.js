import { css } from 'glamor';

export default css({
  padding: 0,
  margin: 0,
  border: 'none',
  textAlign: 'left',
  alignItems: 'stretch',
  width: '100%',
  cursor: 'pointer',
}).toString();
