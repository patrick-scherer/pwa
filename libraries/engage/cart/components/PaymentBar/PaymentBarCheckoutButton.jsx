// @flow
import React, { useContext, useMemo } from 'react';
import I18n from '@shopgate/pwa-common/components/I18n';
import Link from '@shopgate/pwa-common/components/Link';
import { SurroundPortals } from '@shopgate/engage/components';
import { CART_CHECKOUT_BUTTON } from '@shopgate/pwa-common-commerce/cart/constants/Portals';
import RippleButton from '@shopgate/pwa-ui-shared/RippleButton';
import { CHECKOUT_PATH } from '@shopgate/pwa-common/constants/RoutePaths';
import { GUEST_CHECKOUT_PATTERN } from '@shopgate/engage/checkout';
import { CartContext } from '../../cart.context';
import connect from './PaymentBarCheckoutButton.connector';
import { button, disabledButton } from './PaymentBarCheckoutButton.style';

type Props = {
  isOrderable?: boolean;
  isGuestCheckoutActive?: boolean;
}

/**
 * Renders the cart payment bar checkout button.
 * @param {boolean} isOrderable Whether the cart is orderable.
 * @return {JSX}
 */
function PaymentBarCheckoutButton({ isOrderable, isGuestCheckoutActive }: Props) {
  const { isLoading } = useContext(CartContext);
  const isActive = useMemo(() => (isOrderable && !isLoading), [isLoading, isOrderable]);
  const checkoutPath = useMemo(
    () => (!isGuestCheckoutActive ? CHECKOUT_PATH : GUEST_CHECKOUT_PATTERN),
    [isGuestCheckoutActive]
  );

  return (
    <SurroundPortals portalName={CART_CHECKOUT_BUTTON} props={{ isActive }}>
      <Link href={checkoutPath} disabled={!isActive}>
        <RippleButton
          disabled={!isActive}
          type="regular"
          className={isActive ? button : disabledButton}
        >
          <I18n.Text string="cart.checkout" />
        </RippleButton>
      </Link>
    </SurroundPortals>
  );
}

PaymentBarCheckoutButton.defaultProps = {
  isOrderable: true,
  isGuestCheckoutActive: false,
};

export default connect(PaymentBarCheckoutButton);
