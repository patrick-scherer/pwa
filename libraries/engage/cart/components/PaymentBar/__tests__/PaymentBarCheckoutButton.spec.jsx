import React from 'react';
import { shallow } from 'enzyme';
import Link from '@shopgate/pwa-common/components/Link';
import PaymentBarCheckoutButton from '../PaymentBarCheckoutButton';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: () => ({
    isLoading: false,
  }),
}));
jest.mock('@shopgate/engage/components');
jest.mock('../PaymentBarCheckoutButton.connector', () => cmp => cmp);

describe('<PaymentBarCheckoutButton />', () => {
  it('should render disabled button', () => {
    const wrapper = shallow(<PaymentBarCheckoutButton isOrderable={false} />).dive();
    const childLink = wrapper.find(Link);

    expect(wrapper).toMatchSnapshot();
    expect(childLink.props().disabled).toBe(true);
    expect(wrapper.find('Translate').props().string).toBe('cart.checkout');
  });

  it('should render enabled button', () => {
    const wrapper = shallow(<PaymentBarCheckoutButton isOrderable />).dive();
    const childLink = wrapper.find(Link);

    expect(wrapper).toMatchSnapshot();
    expect(childLink.props().disabled).toBe(false);
  });
});
