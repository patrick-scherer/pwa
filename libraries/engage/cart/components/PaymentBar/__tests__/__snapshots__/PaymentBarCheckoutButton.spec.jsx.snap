// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaymentBarCheckoutButton /> should render disabled button 1`] = `
<Fragment>
  <Portal
    name="cart.checkout-button.before"
    props={null}
  />
  <Portal
    name="cart.checkout-button"
    props={null}
  >
    <Connect(Link)
      disabled={true}
      href="/checkout"
    >
      <RippleButton
        className="css-ce5yef"
        disabled={true}
        flat={false}
        onClick={null}
        rippleClassName=""
        rippleSize={null}
        testId="Button"
        type="regular"
        wrapContent={true}
      >
        <Translate
          className={null}
          params={Object {}}
          role={null}
          string="cart.checkout"
          transform={null}
        />
      </RippleButton>
    </Connect(Link)>
  </Portal>
  <Portal
    name="cart.checkout-button.after"
    props={null}
  />
</Fragment>
`;

exports[`<PaymentBarCheckoutButton /> should render enabled button 1`] = `
<Fragment>
  <Portal
    name="cart.checkout-button.before"
    props={null}
  />
  <Portal
    name="cart.checkout-button"
    props={null}
  >
    <Connect(Link)
      disabled={false}
      href="/checkout"
    >
      <RippleButton
        className="css-4uvyh3"
        disabled={false}
        flat={false}
        onClick={null}
        rippleClassName=""
        rippleSize={null}
        testId="Button"
        type="regular"
        wrapContent={true}
      >
        <Translate
          className={null}
          params={Object {}}
          role={null}
          string="cart.checkout"
          transform={null}
        />
      </RippleButton>
    </Connect(Link)>
  </Portal>
  <Portal
    name="cart.checkout-button.after"
    props={null}
  />
</Fragment>
`;
